classes/robot/crawler.php

```diff
@@ -549,53 +551,58 @@ public function process_queue($verbose = false) {
             $recentcourses = $this->get_recentcourses();
         }
 
-        // Iterate through the queue until we find an item that is a recent course, or the time runs out.
+        // Iterate through the queue.
         $cronstart = time();
         $cronstop = $cronstart + $config->maxcrontime;
-        $hasmore = true;
         $hastime = true;
-        while ($hasmore && $hastime) {
-            // Grab the first item from the queue.
-            $node = $DB->get_record_sql('SELECT *
+
+        $timeout = 0; // Lock this resource with a zero second timeout.
+        $locktype = 'tool_crawler_process_queue';
+
+        // Get an instance of the currently configured lock_factory.
+        $lockfactory = \core\lock\lock_config::get_lock_factory($locktype);
+
+        // While we are not exceeding the maxcron time, and the queue is not empty.
+        while ($hastime) {
+            if (empty($nodes)) {
+                // Grab a list of items from the front of the queue
+                $nodes = $DB->get_records_sql('SELECT *
                                          FROM {tool_crawler_url}
                                         WHERE lastcrawled IS NULL
                                            OR lastcrawled < needscrawl
                                      ORDER BY priority DESC, needscrawl ASC, id ASC
-                                        LIMIT 1
-                                    ');
+                                        LIMIT :limitadhocworkers
+                                    ', ['limitadhocworkers' => 1000]);
+                if (empty($nodes)) {
+                    return $hastime; // The queue is empty
```

Comment:
shouldn't this be hard coded to return false?

Reply:
It should return true, since there is still more time but an empty queue. that variable will be true here, but we can hardcode it instead, will make it easier to read.